shunting-yard-replicator/src/main/java/com/hotels/shunting/yard/replicator/exec/event/MetaStoreEvent.java

```diff
@@ -0,0 +1,189 @@
+      if (this.partitionValues == null) {
+        this.partitionValues = new ArrayList<>();
+      }
+      this.partitionValues.add(partitionValues);
+      return this;
+    }
+
+    public Builder parameters(Map<String, String> parameters) {
+      if (this.parameters == null) {
+        this.parameters = new HashMap<>();
+      }
+      if (parameters != null) {
+        this.parameters.putAll(parameters);
+      }
+      return this;
+    }
+
+    public Builder parameter(String key, String value) {
+      if (parameters == null) {
+        parameters = new HashMap<>();
+      }
+      parameters.put(key, value);
+      return this;
+    }
+
+    public Builder deleteData(boolean deleteData) {
+      checkState(isDropEvent(eventType), "");
+      return parameter(DELETE_DATA_PARAMETER, Boolean.toString(deleteData));
+    }
+
+    public Builder environmentContext(Map<String, String> environmentContext) {
+      if (this.environmentContext == null) {
+        this.environmentContext = new HashMap<>();
+      }
+      if (environmentContext != null) {
+        this.environmentContext.putAll(environmentContext);
+      }
+      return this;
+    }
+
+    public MetaStoreEvent build() {
+      return new MetaStoreEvent(this);
+    }
+  }
+
+  public static Builder builder(EventType eventType, String databaseName, String tableName) {
+    return new Builder(eventType, databaseName, tableName);
+  }
+
+  private static boolean isDropEvent(EventType eventType) {
+    return eventType == ON_DROP_PARTITION || eventType == ON_DROP_TABLE;
+  }
+
+  private final EventType eventType;
+  private final String databaseName;
+  private final String tableName;
+  private final List<String> partitionColumns;
+  private final List<List<String>> partitionValues;
+  private final Map<String, String> parameters;
+  private final Map<String, String> environmentContext;
+
+  private MetaStoreEvent(Builder builder) {
+    eventType = builder.eventType;
+    databaseName = builder.databaseName;
+    tableName = builder.tableName;
+    partitionColumns = builder.partitionColumns;
+    partitionValues = builder.partitionValues == null ? null : ImmutableList.copyOf(builder.partitionValues);
+    parameters = builder.parameters == null ? null : ImmutableMap.copyOf(builder.parameters);
+    environmentContext = builder.environmentContext == null ? null : ImmutableMap.copyOf(builder.environmentContext);
+  }
+
+  public EventType getEventType() {
+    return eventType;
+  }
+
+  public String getDatabaseName() {
+    return databaseName;
+  }
+
+  public String getTableName() {
+    return tableName;
+  }
+
+  public List<String> getPartitionColumns() {
+    return partitionColumns;
+  }
+
+  public List<List<String>> getPartitionValues() {
+    return partitionValues;
+  }
+
+  public Map<String, String> getParameters() {
+    return parameters;
+  }
+
+  public Map<String, String> getEnvironmentContext() {
+    return environmentContext;
+  }
+
```

Comment:
This code is duplicated in common and here, maybe refactor?

Reply:
This return the `environmentContext` property of the object, I can't refactor

Reply:
Yeah, we have this joining on the dot in quite a few places, it's a one liner which would turn into calls on some kind of Util object (maybe in common-hive-metastore?) which would be a minor improvement I guess. I'm OK with this for now but something to keep in mind whenever we find ourselves repeating this snippet.

Reply:
I see, the snapshot I get is from a different method. I wouldn't bother with this because the logic is really simple and won't ever change, to make it a util method then we need to either define an interface or pass the 2 values to the util method which doesn't make much difference but adds more boilerplate.